File: real-estate-web-be/web_microservice_be/ListingService/src/main/java/com/example/ListingService/service/ListingService.java
package com.example.ListingService.service;

import com.example.ListingService.request.PropertyActionRequest;
import com.example.ListingService.request.CreatePropertyRequest;
import com.example.ListingService.request.UpdatePropertyRequest;
import com.example.ListingService.model.Property;
import com.example.ListingService.response.PropertyResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ListingService {

    // --- USER FEATURE ---

    // Create new property(draft)
    PropertyResponse createProperty(Long userId, CreatePropertyRequest request);

    // Submit property to review
    void submitProperty(Long userId, Long propertyId);

    // Update property
    PropertyResponse updateProperty(Long userId, Long propertyId, UpdatePropertyRequest request);

    // Change visibility of property
    void toggleVisibility(Long userId, Long propertyId);

    // Mark property as sold
    void markAsSold(Long userId, Long propertyId);

    // Mark property as rented
    void markAsRented(Long userId, Long propertyId);

    // Delete property (soft delete)
    void deleteProperty(Long userId, Long propertyId);

    // Get my listings with filters
    Page<PropertyResponse> getMyListings(Long userId, String search, Property.Status status, Pageable pageable);

    // Get my sold/rent listings with filters
    Page<PropertyResponse> getMySoldRentListings(Long userId, String search, Property.Status status, Pageable pageable);

    // Get my listing detail
    PropertyResponse getMyListingDetail(Long userId, Long propertyId);

    // --- ADMIN FEATURE ---

    // Review and approve/reject property
    void approveProperty(Long propertyId, PropertyActionRequest request);

    // Get pending listings for review
    Page<PropertyResponse> getPendingListings(String search, Pageable pageable);

    // Get user listings by admin with filters
    Page<PropertyResponse> getUserListingsByAdmin(Long userId, String search, Property.Status status, boolean includeDeleted, Pageable pageable);

    // Get listing detail by admin
    PropertyResponse getListingDetailByAdmin(Long propertyId);

    // --- PUBLIC FEATURE (KHÔNG CẦN LOGIN) ---

    // Lấy danh sách bài đăng khả dụng (AVAILABLE)
    Page<PropertyResponse> getPublicListings(String search,
                                     String propertyTransactionType,
                                     String propertyType,
                                     Double minPrice, Double maxPrice,
                                     Pageable pageable);

    // Xem chi tiết bài đăng (chỉ bài AVAILABLE)
    PropertyResponse getPublicListingDetail(Long propertyId);
}

File: real-estate-web-fe/src/services/listingService .js
import http from './http';
export const listingService = {
    // /public
    getAllListing: async (search, transactionType, propertyType, minPrice, maxPrice, page = 0, size = 10) => {
        const params = {
            search,
            transactionType,
            propertyType,
            minPrice,
            maxPrice,
            page,
            size,
        };
        console.log("params", params);

        const res = await http.get(`/listing/public`, { params });
        console.log("data lis", res);
        return res.data;
    },
    // /property/{id}
    // /property/{id}/submit
    submitDraftListing: async (id) => {
        const res = await http.post(`/listing/property/${id}/submit`);
        return res.data;
    },
    // /property/{id}/toggle-visibility
    hiddenListing: async (id) => {
        const res = await http.post(`/listing/property/${id}/toggle-visibility`);
        return res.data;
    },
    // /property/{id}/sold
    hasSoldListing: async (id) => {
        const res = await http.post(`/listing/property/${id}/sold`);
        return res.data;
    },
    // /property/{id}/rented
    hasRentedListing: async (id) => {
        const res = await http.post(`/listing/property/${id}/rented`);
        return res.data;
    },
    // /property/{id}
    updateListing: async (id, data) => {
        const res = await http.put(`/listing/property/${id}`, data);
        return res.data;
    },
    // /my-listings
    getUserListing: async (status, page = 0, size = 10) => {
        const params = {
            status,
            page,
            size,
        };
        const res = await http.get(`/listing/my-listings`, { params });
        console.log("user lis", res.data);

        return res.data;
    },
    // /my-history
    transferHistoryListing: async (status, page = 0, size = 10) => {
        const params = {
            status,
            page,
            size,
        };
        const res = await http.get(`/listing/my-history`, { params });
        return res.data;
    },
    // /property/{id}
    deleteListing: async (id) => {
        const res = await http.delete(`/listing/property/${id}`);
        return res.data;
    },
    // /admin/property/pending
    getUserListingByAdmin: async (status, page = 0, size = 10) => {
        const params = {
            status,
            page,
            size,
        };
        const res = await http.get(`/listing/admin/properties`, { params });
        return res.data;
    },
    // /admin/user/{userId}/properties
    getUserAllUserListingByAdmin: async (userId, page = 0, size = 10) => {
        const params = {
            includeDeleted: true,
            page,
            size,
        };
        const res = await http.get(`/listing/admin/user/${userId}/properties`, { params });
        return res.data;
    },
    // /admin/property/{id}
    // /admin/property/{id}/approve
    approveListing: async (id, data) => {
        const res = await http.post(`/listing/admin/property/${id}/approve`, data);
        return res.data;
    },
    // /public/{id}
    getListingDetails: async (id) => {
        const res = await http.get(`/listing/public/${id}`);
        return res.data.data;
    },
    // /property
    createListing: async (data) => {
        const res = await http.post(`/listing/property`, data);
        return res.data;
    }
}